src/Microsoft.AspNet.OData.Shared/ODataAPIHandler.cs

```diff
@@ -64,129 +69,245 @@ public abstract class ODataAPIHandler<TStructuralType>: IODataAPIHandler where T
                 {
-                    object propVal = prop.GetValue(obj);
-                    if (propVal == null)
-                    {
-                        continue;
-                    }
+                    continue;
+                }
+
+                object parentObj = GetObjectWithoutNavigationProperties(obj, navPropNames);
 
-                    if (propVal is IEnumerable lst)
+                object[] nestedHandlerParams = new object[] { parentObj, navPropertName };
+                IODataAPIHandler nestedHandler = (IODataAPIHandler)apiHandler.GetType().GetMethod(nameof(GetNestedHandler)).Invoke(apiHandler, nestedHandlerParams);
+
+                if (navPropVal is IEnumerable lst)
+                {
+                    foreach (var item in lst)
                     {
-                        foreach (object item in lst)
+                        if (item.GetType().IsPrimitive)
                         {
-                            if (item.GetType().IsPrimitive)
-                            {
-                                break;
-                            }
-
-                            CheckAndApplyODataId(item, model);
+                            break;
                         }
-                    }
-                    else
-                    {
-                        CheckAndApplyODataId(propVal, model);
+
+                        CopyObjectProperties(item, model, nestedHandler, apiHandlerFactory);
                     }
                 }
+                else
+                {
+                    CopyObjectProperties(navPropVal, model, nestedHandler, apiHandlerFactory);
+                }
             }
         }
 
-        private object ApplyODataIdOnContainer(object obj, ODataPath odataPath)
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "We don't have a specific exception to catch.")]
+        internal static void ApplyHandlers(IODataAPIHandler odataApiHandler, IODataAPIHandler odataIdContainerHandler, object resource, IDictionary<string, object> keys, List<string> navigationProperties, out bool failedOperation)
         {
-            KeySegment keySegment = odataPath.LastOrDefault() as KeySegment;
-            Dictionary<string, object> keys = keySegment?.Keys.ToDictionary(x => x.Key, x => x.Value);
-            if (keys != null)
+            failedOperation = false;
+            DataModificationOperationKind operation = DataModificationOperationKind.Insert;
+
+            try
             {
-                TStructuralType returnedObject;
-                string error;
-                if (this.ParentObject.Equals(obj))
+                object[] handlerParams = new object[] { keys, null, null };
+
+                IODataAPIHandler handlerForGet = odataApiHandler;
+                ODataAPIResponseStatus getResponse = ODataAPIResponseStatus.NotFound;
+                object getObject = null;
```

Comment:
```suggestion
                object returnedObject= null;
```
`getObject` is a bit confusing here.